### PR TITLE
Add support for Python 3.11

### PR DESCRIPTION
**Pull request by hugovk**
Python 3.11 was [released on 2022-10-24](https://discuss.python.org/t/python-3-11-0-final-is-now-available/20291?u=hugovk) 🚀 

[image](https://discuss.python.org/t/python-3-11-0-final-is-now-available/20291?u=hugovk)

Also test Python 3.12-dev and drop Python 3.6, end-of-life since 2021-12-23.

And we can use the builtin caching of actions/setup-python to remove some config: https://github.com/actions/setup-python#caching-packages-dependencies

### CODE REVIEW

**Review by thombashi — CHANGES_REQUESTED** [review]
I would like to keep Python 3.6 support for a while.
As you mentioned, Python 3.6 is already reached the EOL.
However, some users of my libraries that depend on this package still use Python 3.6
If this package kept Python 3.6 support, such users could also enjoy performance improvement.

**Comment by hugovk** [comment]
Sure, updated to keep Python 3.6.

GitHub Actions is dropping Python 3.6 too and it isn't available on `ubuntu-latest` aka `ubuntu-22.04`, so we'll need to stick with `ubuntu-20.04`.

File: pyproject.toml [hunk 1]
```diff
@@ -18,7 +18,7 @@ exclude = '''
 )/
 | docs/conf.py
 '''
-target-version = ['py36', 'py37', 'py38', 'py39', 'py310']
+target-version = ['py37', 'py38', 'py39', 'py310', 'py311']
```
Comment (thombashi) [inline pyproject.toml:21]: Could you revert Python 3.6 part for this?

File: pyproject.toml [hunk 2]
```diff
@@ -56,7 +56,7 @@ skip_glob = [
 
 [tool.mypy]
 ignore_missing_imports = true
-python_version = 3.6
+python_version = 3.7
```
Comment (thombashi) [inline pyproject.toml:59]: Same here

**Review by thombashi — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (hugovk) [reply pyproject.toml]: Thanks, reverted this and the other.

**Review by thombashi — APPROVED** [review]
LGTM
Thank you for your pull request.

**Pull request merged**